pkg/blockbuilder/blockbuilder.go

```diff
@@ -542,6 +543,39 @@ func (b *BlockBuilder) consumePartitionSection(
 
 	level.Info(logger).Log("msg", "start consuming")
 
+	var (
+		done     = make(chan struct{})
+		fetchesC = make(chan kgo.Fetches, 100)
+		wg       sync.WaitGroup
+	)
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			select {
+			case <-done:
+				return
+			case <-ctx.Done():
+				return
+			default:
+				// Put a timeout on PollFetches to avoid indefinite blocking when we reach the
+				// end of partitions.
+				pollCtx, cancel := context.WithTimeout(ctx, 2*time.Second)
+				fetches := b.kafkaClient.PollFetches(pollCtx)
```

Comment:
`PollFetches` doesn't actually do any work. It waits until the client receives records from kafka. And the client only has one inflight Fetch request per partition. The client waits for that Fetch request to finish before firing off the next one. It does them sequentially because it uses the offset of the last record in the first response as the offset in the second request. That's the reason we implemented our own concurrentFetchers